### PR TITLE
user group create

### PR DESCRIPTION
**Pull request by teamdraftbox**
feature create group form created

### CODE REVIEW

File: src/components/pages/group/form/speciesCheckBox/adminInviteField/index.tsx [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+  label?: string;
+  textBoxName: string;
+  textBoxLabel?: string;
+  fieldGroupTitle: string;
+  isMultiple?: boolean;
+}
+
+const AdminInviteField = (props: AdminListDeatils) => {
+  const { t } = useTranslation();
+  const onAdminUserQuery = async (q) => {
```
Comment (harshzalavadiya) [inline src/components/pages/group/form/speciesCheckBox/adminInviteField/index.tsx:30]: Please format code properly leave space after each function for better readability

File: src/components/pages/group/form/speciesCheckBox/adminInviteField/index.tsx [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    const { data } = await axUserSearch(q);
+    return data.map((tag) => ({ label: tag.name, value: tag.id, version: tag.version }));
+  };
+  return (
+    <div>
+      <h1 className="mt4" style={{ margin: "20px 15px", fontSize: 25 }}>
```
Comment (harshzalavadiya) [inline src/components/pages/group/form/speciesCheckBox/adminInviteField/index.tsx]: Use Heading component here instead of plainly using h1 and inline styles

File: src/components/pages/group/form/speciesCheckBox/adminInviteField/index.tsx [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+        <strong>{t(props.adminTitle)}</strong>
+      </h1>
+      {props.adminList &&
+        props.adminList.length &&
+        props.adminList.map((item: AdminDeatils, index) => (
+          <div className="white-box-padding" key={index}>
```
Comment (harshzalavadiya) [inline src/components/pages/group/form/speciesCheckBox/adminInviteField/index.tsx]: This might be uncesessary please use chakra UI's style props and use `Box` component https://chakra-ui.com/style-props

File: src/components/pages/group/form/speciesCheckBox/adminInviteField/index.tsx [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+              <h3>{item.fieldGroupTitle}</h3>
+            </div>
+            <div>
+              <SelectAsync
+                placeholder={t(`GROUP.ADMIN_INVITE`)}
+                name={item.name}
+                form={props.form}
+                onQuery={onAdminUserQuery}
+                multiple={item.isMultiple ? item.isMultiple : false}
+                label={item.label ? item.label : ""}
```
Comment (harshzalavadiya) [inline src/components/pages/group/form/speciesCheckBox/adminInviteField/index.tsx]: preferablly just pass prop there's no need for turnary condition at all `item.label ? item.label : ""`

File: src/components/pages/group/form/speciesCheckBox/customCheckbox/index.tsx [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+
+  return (
+    <>
+      <FormControl isInvalid={form.errors[name] && true} isRequired={true} mb={mb} {...props}>
+        <FormLabel htmlFor={name}>{label}</FormLabel>
+        {/* <CheckboxGroup id={name} value={value} onChange={onChange} isInline>
+            {options.map((o) => (
+              <SpeciesMultiSelectCheckbox key={o.id} value={o.id} icon={o.name} />
+            ))}
+          </CheckboxGroup> */}
```
Comment (harshzalavadiya) [inline src/components/pages/group/form/speciesCheckBox/customCheckbox/index.tsx]: Please don't comment commented code and remove it everywhere

File: src/components/pages/group/form/speciesCheckBox/customCheckbox/multiple-categorical/checkbox.tsx [hunk 6]
```diff
@@ -0,0 +1,16 @@
+import { VisuallyHidden, Image } from "@chakra-ui/core";
+import React from "react";
+import { getSpeciesIcon } from "@utils/media";
+import Tooltip from "@components/@core/tooltip";
+const VH: any = VisuallyHidden;
+
+const Radio = ({ value, label, icon, ...props }) => (
```
Comment (harshzalavadiya) [inline src/components/pages/group/form/speciesCheckBox/customCheckbox/multiple-categorical/checkbox.tsx]: Please rename class to `Checkbox` it's not `Radio`

File: src/components/pages/group/form/speciesCheckBox/index.tsx [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+    const moderator = getAdminUser(group.moderator);
+    const payload = {
+      allowUserToJoin: group.group_invitaion,
+      description: group.group_description,
+      icon: group.group_icon,
+      name: group.group_name,
+      neLatitude: group && group.spacial_coverage.ne ? group.spacial_coverage.ne[0] : "",
+      neLongitude: group && group.spacial_coverage.ne ? group.spacial_coverage.ne[1] : "",
+      swLatitude: group && group.spacial_coverage.se ? group.spacial_coverage.se[0] : "",
+      swLongitude: group && group.spacial_coverage.se ? group.spacial_coverage.se[1] : "",
```
Comment (harshzalavadiya) [inline src/components/pages/group/form/speciesCheckBox/index.tsx]: Could have been written in a simple manner, use `Optional Chaining + Destructure` here instead of this

File: src/components/pages/group/form/speciesCheckBox/index.tsx [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+      languageId: 0,
+      spacialCOverage: group.sGroup ? group.sGroup : [],
+      habitatCoverage: group.habitatCoverage ? group.habitatCoverage : [],
```
Comment (harshzalavadiya) [inline src/components/pages/group/form/speciesCheckBox/index.tsx]: There's no need to add condition here this could have been `group.habitatCoverage || []` as well

or preferabally name keys appropriately and have a default value so you might have avoided this transformation

File: src/components/pages/group/form/speciesCheckBox/index.tsx [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+            form={hform}
+          />
+          <RadioInputField
+            form={hform}
+            isInline={false}
+            name="group_invitaion"
+            label="Can user join the group without invitaion"
+            options={[
+              { label: "yes", value: "true" },
+              { label: "no", value: "false" }
```
Comment (harshzalavadiya) [inline src/components/pages/group/form/speciesCheckBox/index.tsx]: Please export options into seprate files or use existing one if there's already one

File: src/components/pages/group/form/speciesCheckBox/index.tsx [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+            name="habitatCoverage"
+            label={t("GROUP.HABITATS_COVERED")}
+            options={habitats}
+            form={hform}
+          />
+          <RadioInputField
+            form={hform}
+            isInline={false}
+            name="group_invitaion"
+            label="Can user join the group without invitaion"
```
Comment (harshzalavadiya) [inline src/components/pages/group/form/speciesCheckBox/index.tsx]: move this text to translation file

File: src/components/pages/group/form/speciesCheckBox/index.tsx [hunk 11]
```diff
@@ -0,0 +1,167 @@
...
+              form={hform}
+              adminTitle={"GROUP.GROUP_ADMIN_TITLE"}
+            />
+            <Button
+              className="white-box-padding"
+              type="submit"
+              onSubmit={handleFormSubmit}
+              variantColor="green"
+            >
+              Button
```
Comment (harshzalavadiya) [inline src/components/pages/group/form/speciesCheckBox/index.tsx]: what's this?

File: src/components/pages/group/form/speciesCheckBox/index.tsx [hunk 12]
```diff
@@ -0,0 +1,167 @@
...
+            </Button>
+          </div>
+        </form>
+      </Box>
+    </Box>
+  ) : (
+    <SavingGroup />
```
Comment (harshzalavadiya) [inline src/components/pages/group/form/speciesCheckBox/index.tsx]: Use submit button's loading state instead of hiding entire form because this might be better user experiance in this scenario

File: src/components/pages/group/form/uploader/imageCard.tsx [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+  getFallbackByMIME,
+  getMyUploadsThumbnail,
+  getObservationThumbnail,
+  getYoutubeImage
+} from "@utils/media";
+import { useStoreState } from "easy-peasy";
+import React, { useMemo } from "react";
+
+import StatusIcon from "@components/pages/observation/create/form/uploader/statusicon";
+import useObservationCreate from "@components/pages/observation/create/form/uploader/use-observation-resources";
```
Comment (harshzalavadiya) [inline src/components/pages/group/form/uploader/imageCard.tsx]: Please sort imports properly
use something like this for better DX https://marketplace.visualstudio.com/items?itemName=rbbit.typescript-hero

File: src/components/pages/group/form/uploader/imageCard.tsx [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+          src={imageURL}
+        />
+        <StatusIcon type={resource.status} />
+      </ImageBox>
+      <CloseButton
+        position="absolute"
+        top="200"
+        right="200"
+        size="lg"
+        color="#FF0000"
```
Comment (harshzalavadiya) [inline src/components/pages/group/form/uploader/imageCard.tsx:60]: please avoid applying hex colors directly we already have selection of colors to choose from

File: src/components/pages/group/form/uploader/index.tsx [hunk 15]
```diff
@@ -0,0 +1,59 @@
...
+import useObservationCreate from "@components/pages/observation/create/form/uploader/use-observation-resources";
+import ResourceCard from "./imageCard";
+import { ObservationCreateProvider } from "@components/pages/observation/create/form/uploader/use-observation-resources";
+import { FormContextValues } from "react-hook-form";
+import { DB_CONFIG } from "@static/observation-create";
+import IndexedDBProvider from "use-indexeddb";
+const DropTarget = dynamic(
+  import("../../../observation/create/form/uploader/observation-resources/drop-target"),
+  { ssr: false }
+);
```
Comment (harshzalavadiya) [inline src/components/pages/group/form/uploader/index.tsx]: use `@obervation` to import files from

File: src/components/pages/group/form/uploader/index.tsx [hunk 16]
```diff
@@ -0,0 +1,59 @@
...
+export default function GroupIconUploader(props: IDropzoneProps) {
+  const { name, form } = props;
+  return (
+    <IndexedDBProvider config={DB_CONFIG} loading="Loading...">
+      <ObservationCreateProvider
+        observationAssets={props.form.control.defaultValuesRef.current[props.name]}
+      >
+        <IconUploader name={name} form={form} />
+      </ObservationCreateProvider>
+    </IndexedDBProvider>
```
Comment (harshzalavadiya) [inline src/components/pages/group/form/uploader/index.tsx]: No need to involve indexeddb in this scenario at all

File: src/components/pages/group/saving/saving.tsx [hunk 17]
```diff
@@ -0,0 +1,179 @@
+import React from "react";
```
Comment (harshzalavadiya) [inline src/components/pages/group/saving/saving.tsx]: No need of this file at all

File: src/components/pages/observation/create/form/uploader/observation-resources/drop-target.tsx [hunk 18]
```diff
@@ -11,7 +11,7 @@ const DropTargetBox = styled.div`
   border: 2px dashed var(--gray-300);
   border-radius: 0.5rem;
   padding: 1rem;
-  min-height: 22rem;
+  min-height: 10rem;
```
Comment (harshzalavadiya) [inline src/components/pages/observation/create/form/uploader/observation-resources/drop-target.tsx:14]: Why this change was made?

File: src/i18n/en.json [hunk 19]
```diff
@@ -501,5 +501,26 @@
...
+    "CREATING": "We're creating your group",
+    "SUCCESS": "Created successfully",
+    "GROUP_ICON": "Icon",
+    "SPECIES_GROUP_COVERAGE": "Species Group Coverage",
+    "HABITATS_COVERED": "Habitats Covered",
+    "SPATIAL_COVERGE": "Spatial Coverage",
+    "TAGS": "Tags",
+    "GROUP_ADMIN_TITLE": "Administrators",
+    "GROUP_INVITE_FOUNDER": "Invite Founders",
+    "GROUP_INVITE_MODERATORS": "Invite Moderators",
```
Comment (harshzalavadiya) [inline src/i18n/en.json]: Please name keys properly for example there's no need to prefix keys with group if they already are under group

File: src/pages/group/create/index.tsx [hunk 20]
```diff
@@ -0,0 +1,44 @@
...
+  const { push, asPath } = useLocalRouter();
+
+  const init = async () => {
+    const user = getNookie(TOKEN.USER);
+    if (user) {
+      setIsLoggedIn(true);
+    } else {
+      push("/login", true, { forward: encode(asPath) });
+    }
+  };
```
Comment (harshzalavadiya) [inline src/pages/group/create/index.tsx]: Make role check in `getServerSide` props and use `hasAccess` instead

File: src/services/usergroup.service.ts [hunk 21]
```diff
@@ -48,3 +48,15 @@ export const axGetPages = async () => {
     return { success: false, data: [] };
   }
 };
+
+export const axCreateGroup = async ({ ...payload }) => {
+  try {
+    const { data } = await http.post(`${ENDPOINT.USERGROUP}/v1/group/create`, {
+      ...payload
```
Comment (harshzalavadiya) [inline src/services/usergroup.service.ts]: there's no need to spread here you can just use `payload`

File: src/services/usergroup.service.ts [hunk 22]
```diff
@@ -48,3 +48,15 @@ export const axGetPages = async () => {
     return { success: false, data: [] };
   }
 };
+
+export const axCreateGroup = async ({ ...payload }) => {
```
Comment (harshzalavadiya) [inline src/services/usergroup.service.ts]: instead of collecting destructuring and spread it's redundant just use `async (payload) => ...`

File: src/services/utility.service.ts [hunk 23]
```diff
@@ -37,3 +37,16 @@ export const axGetYouTubeInfo = async (url) => {
...
   }
 };
+
+export const axGetAllHabitat = async () => {
+  try {
+    const { data } = await axios.get(`${ENDPOINT.UTILITY}/v1/services/habitat/all`);
+    return {
+      success: true,
+      data
+    };
```
Comment (harshzalavadiya) [inline src/services/utility.service.ts]: This could have been single line

File: src/static/events.ts [hunk 24]
```diff
@@ -6,6 +6,9 @@ export const RESOURCES_UPLOADING = "resources_uploading";
 export const SYNC_SINGLE_OBSERVATION = "sync_single_observation";
 export const SYNC_SINGLE_OBSERVATION_DONE = "sync_single_observation_done";
 export const SYNC_SINGLE_OBSERVATION_ERROR = "sync_single_observation_error";
+export const SYNC_SINGLE_GROUP = "sync_signle_group";
+export const SYNC_SINGLE_GROUP_DONE = "sync_single_group_done";
+export const SYNC_SINGLE_GROUP_ERROR = "sync_signle_group_error";
```
Comment (harshzalavadiya) [inline src/static/events.ts]: No need of this since there's no offline support needed

**Review by harshzalavadiya — CHANGES_REQUESTED** [review]

File: src/static/events.ts (hunk 24 above)
Comment (teamdraftbox) [reply src/static/events.ts]: sure ill remove the online support code

File: src/pages/group/create/index.tsx [hunk 25]
```diff
@@ -0,0 +1,35 @@
...
+
+export async function getServerSideProps(ctx) {
+  if (!hasAccess([Role.Admin, Role.User], ctx)) {
+    const Location = `/login?forward=${encode("/group/create")}`;
+    ctx.res.writeHead(302, {
+      Location,
+      "Content-Type": "text/html; charset=utf-8"
+    });
+    ctx.res.end();
+  }
```
Comment (harshzalavadiya) [inline src/pages/group/create/index.tsx]: `authorizedPageSSR` maybe?

**Review by harshzalavadiya — CHANGES_REQUESTED** [review]

**Comment by harshzalavadiya** [comment]
### Bug

- If user is not logged in and hits `/group/create` page throws error

File: src/components/pages/group/edit/index.tsx [hunk 26]
```diff
@@ -0,0 +1,41 @@
...
+  const [userGroup, setUserGroup] = useState();
+  const [adminMembers, setAdminMembers] = useState();
+
+  useEffect(() => {
+    axGetGroupEditById(id)
+      .then(({ success, data }) => {
+        if (success) setUserGroup(data);
+      })
+      .catch((err) => {
+        err;
```
Comment (harshzalavadiya) [inline src/components/pages/group/edit/index.tsx]: throw a console.error here

File: src/components/pages/group/edit/index.tsx [hunk 27]
```diff
@@ -0,0 +1,41 @@
...
+      });
+    axGetGroupMembersById(id)
+      .then(({ success, data }) => {
+        if (success) setAdminMembers(data);
+      })
+      .catch((err) => err);
+  }, []);
+
+  if (userGroup && adminMembers) {
```
Comment (harshzalavadiya) [inline src/components/pages/group/edit/index.tsx]: add a return and put a turnery operator

File: src/components/pages/group/edit/userEditForm.tsx [hunk 28]
```diff
@@ -0,0 +1,233 @@
...
+          variant="ghost"
+          onClick={() => {
+            setShow(!show);
+          }}
+          justifyContent="space-between"
+          width="100%"
+          size="lg"
+        >
+          {t("GROUP.ADMIN_TITLE")}
+          {<Icon name={show ? "minus" : "add"} />}
```
Comment (harshzalavadiya) [inline src/components/pages/group/edit/userEditForm.tsx]: Unnecessary Brackets

File: src/components/pages/group/form/speciesCheckBox/adminInviteField/index.tsx [hunk 29]
```diff
@@ -0,0 +1,90 @@
...
+    const { data } = await axUserSearch(q);
+    return data.map((tag) => ({ label: tag.name, value: tag.id, version: tag.version }));
+  };
+
+  const handleDelete = async (userId, name) => {
+    if (
+      adminMember &&
+      adminMember[name] &&
+      adminMember[name].some((item) => item.value == userId?.value)
+    ) {
```
Comment (harshzalavadiya) [inline src/components/pages/group/form/speciesCheckBox/adminInviteField/index.tsx:40]: use optional chaining here

File: src/components/pages/group/form/speciesCheckBox/adminInviteField/index.tsx [hunk 30]
```diff
@@ -0,0 +1,90 @@
...
+  return (
+    <div>
+      {props.adminTitle && (
+        <Heading as="h2" className="mt4" m="20px" fontSize="x-large">
+          Administrator
+        </Heading>
+      )}
+      {props.adminList &&
+        props.adminList.length &&
+        props.adminList.map((item: AdminDeatils, index) => (
```
Comment (harshzalavadiya) [inline src/components/pages/group/form/speciesCheckBox/adminInviteField/index.tsx]: Optional chaining

File: src/components/pages/group/form/speciesCheckBox/adminInviteField/index.tsx [hunk 31]
```diff
@@ -0,0 +1,90 @@
...
+          <Box m={15} key={index}>
+            <div>
+              <h3>{item.fieldGroupTitle}</h3>
```
Comment (harshzalavadiya) [inline src/components/pages/group/form/speciesCheckBox/adminInviteField/index.tsx]: use <PageHeading/>?

File: src/components/pages/group/form/speciesCheckBox/customCheckbox/index.tsx [hunk 32]
```diff
@@ -0,0 +1,47 @@
+import { Divider, FormControl, FormErrorMessage, FormHelperText, FormLabel } from "@chakra-ui/core";
```
Comment (harshzalavadiya) [inline src/components/pages/group/form/speciesCheckBox/customCheckbox/index.tsx:1]: Modify Existing one to support multiple selection? or just convert it to dropdown multi select component

File: src/components/pages/group/form/uploader/IconUploader.tsx [hunk 33]
```diff
@@ -0,0 +1,91 @@
...
+    display: block;
+    font-size: 2.5rem;
+    margin: 0 auto;
+    margin-bottom: 0.5rem;
+  }
+`;
+
+export const accept = ["image/jpg", "image/jpeg", "image/png", "video/*", "audio/*"];
+
+export default function DropTarget({ assetsSize, setValue }) {
```
Comment (harshzalavadiya) [inline src/components/pages/group/form/uploader/IconUploader.tsx]: No need for entire dropzone here just use simple file with thumbnail preview?

File: src/components/pages/group/form/uploader/index.tsx [hunk 34]
```diff
@@ -0,0 +1,46 @@
+import { SimpleGrid } from "@chakra-ui/core";
+import dynamic from "next/dynamic";
+import React, { useEffect, useState } from "react";
+import useObservationCreate from "@components/pages/observation/create/form/uploader/use-observation-resources";
+import ResourceCard from "./imageCard";
+import { FormContextValues } from "react-hook-form";
+const DropTarget = dynamic(import("./IconUploader"), { ssr: false });
+export interface IDropzoneProps {
```
Comment (harshzalavadiya) [inline src/components/pages/group/form/uploader/index.tsx:8]: spaces!!

File: src/components/pages/observation/list/header/index.tsx [hunk 35]
```diff
@@ -50,9 +51,11 @@ export default function ListHeader() {
           </Text>
         )}
       </div>
-      <div>
+      <Box display="flex" p={[6]}>
```
Comment (harshzalavadiya) [inline src/components/pages/observation/list/header/index.tsx]: No arrays just number also works

File: src/components/pages/observation/list/views/shared/filterDownLoadModal.tsx [hunk 36]
```diff
@@ -0,0 +1,120 @@
...
+import { useStoreState } from "easy-peasy";
+import { axObservationFilterDownload } from "@services/observation.service";
+import notification, { NotificationType } from "@utils/notification";
+import * as Yup from "yup";
+
+export default function DownLoadFilterModal({ traits, customFields }) {
+  const { isOpen, onOpen, onClose } = useDisclosure();
+  const { id } = useStoreState((s) => s.user);
+  const { t } = useTranslation();
+  const hform = useForm({
```
Comment (harshzalavadiya) [inline src/components/pages/observation/list/views/shared/filterDownLoadModal.tsx:34]: Hook form is just an overkill here, plus if you are not going to use fields just merge it when submitting request

File: src/pages/observation/list.tsx [hunk 37]
```diff
@@ -32,7 +43,14 @@ ObservationListPage.getInitialProps = async (ctx) => {
   )?.id;
 
   const initialFilterParams = { ...DEFAULT_FILTER, ...ctx.query, userGroupList };
+  const fields = ["customFields", "traits"];
   const { data } = await axGetListData(initialFilterParams);
+  const { data: observationList } = await axGetObservationListConfig();
+  fields.forEach((item) => {
+    observationList[item] = observationList[item].map((item) => {
+      return { value: item.name, name: item.name.toLowerCase() };
```
Comment (harshzalavadiya) [inline src/pages/observation/list.tsx]: can avoid return here

File: src/services/observation.service.ts [hunk 38]
```diff
@@ -381,6 +381,14 @@ export const axGetObservationsByResources = async (resources) => {
     return { success: false, data: [] };
   }
 };
+export const axObservationFilterDownload = async (index, type, params) => {
```
Comment (harshzalavadiya) [inline src/services/observation.service.ts:385]: Spaces!!!!

**Review by harshzalavadiya — CHANGES_REQUESTED** [review]
Hi @teamdraftbox Please go through each comment and make changes

**Comment by harshzalavadiya** [comment]
Closing in favor of #6
